chain/ethereum/src/network_indexer/network_indexer.rs

```diff
@@ -0,0 +1,1540 @@
+        transition!(PollChainHead {
+            local_head,
+            prev_chain_head: None,
+            chain_head: poll_chain_head(context),
+        })
+    }
+
+    fn poll_poll_chain_head<'a, 'c>(
+        state: &'a mut RentToOwn<'a, PollChainHead>,
+        context: &'c mut RentToOwn<'c, Context>,
+    ) -> Poll<AfterPollChainHead, Error> {
+        // Abort if the output stream has been closed.
+        try_ready!(context.event_sink.poll_ready());
+
+        match state.chain_head.poll() {
+            // Wait until we have the chain head block.
+            Ok(Async::NotReady) => Ok(Async::NotReady),
+
+            // We have a (new?) chain head, decide what to do.
+            Ok(Async::Ready(chain_head)) => {
+                // Validate the chain head.
+                if chain_head.number.is_none() || chain_head.hash.is_none() {
+                    // This is fairly irregular, so log a warning.
+                    warn!(
+                        context.logger,
+                        "Chain head block number or hash missing; try again";
+                        "block" => format_light_block(&chain_head),
+                    );
+
+                    // Chain head was invalid, try getting a better one.
+                    transition!(PollChainHead {
+                        local_head: state.local_head,
+                        prev_chain_head: state.prev_chain_head,
+                        chain_head: poll_chain_head(context,),
+                    })
+                }
+
+                let state = state.take();
+
+                if Some((&chain_head).into()) != state.prev_chain_head {
+                    context
+                        .metrics
+                        .last_new_chain_head_time
+                        .set(Utc::now().timestamp() as f64)
+                }
+
+                update_chain_and_local_head_metrics(context, &chain_head, state.local_head);
+
+                debug!(
+                    context.logger,
+                    "Identify next blocks to index";
+                    "chain_head" => format_light_block(&chain_head),
+                    "local_head" => state.local_head.map_or(
+                        String::from("none"), |ptr| format_block_pointer(&ptr)
+                    ),
+                );
+
+                // If we're already at the chain head, keep polling it.
+                if Some((&chain_head).into()) == state.local_head {
+                    debug!(
+                        context.logger,
+                        "Already at chain head; poll chain head again";
+                        "chain_head" => format_light_block(&chain_head),
+                        "local_head" => state.local_head.map_or(
+                            String::from("none"), |ptr| format_block_pointer(&ptr)
+                        ),
+                    );
+
+                    // Chain head wasn't new, try getting a new one.
+                    transition!(PollChainHead {
```

Comment:
I thought so too initially; then I ran it and I noticed that — once it has caught up with the chain — it's only generating about 2-3 requests per second (because `eth_getBlockNumber("latest")` takes a few hundred milliseconds usually), which is not excessive. For chains that have block times of 5s or so I wouldn't want to add a delay greater than 500ms. But that would only reduce the requests per second from 2-3 to 1-2. Not sure it's worth it.

Reply:
Alright we can see how this works out in practice.